File: src/engine.js
'use strict';
// Map a Claude Code tool call to capabilities, then decide against a profile.
// capabilityFor turns a PreToolUse envelope into candidate capability strings;
// resolve matches them against a profile under the active mode, returning
// allow / ask / deny with a reason.
const { SECRET_PATH } = require('./patterns');
const { alwaysDenied } = require('./profiles');
const { shellCapabilities } = require('./shellParse');

const PROD_PATH = /(^|\/)(prod|production|infra|terraform|k8s|kubernetes|helm)(\/|$)|(^|\/)(Dockerfile|docker-compose\.ya?ml)$/i;
const MIGRATION_PATH = /(^|\/)(migrations?)(\/|$)|\/migrate\//i;
const AUTH_PATH = /(^|\/)(auth|permissions?|security|crypto)(\/|$)/i;
const BILLING_PATH = /(^|\/)(billing|payments?|stripe)(\/|$)/i;
const CI_PATH = /(^|\/)\.github\/workflows\/|(^|\/)\.gitlab-ci\.ya?ml$|(^|\/)\.circleci\//i;

function writeCapsForPath(path) {
  if (SECRET_PATH.test(path)) return ['secrets:write', 'files:write'];
  if (MIGRATION_PATH.test(path)) return ['migrations:write', 'files:write'];
  if (CI_PATH.test(path)) return ['ci:write', 'files:write'];
  if (AUTH_PATH.test(path)) return ['auth:write', 'files:write'];
  if (BILLING_PATH.test(path)) return ['billing:write', 'files:write'];
  if (PROD_PATH.test(path)) return ['prod:write', 'files:write'];
  return ['files:write'];
}

// Returns [capabilityCandidates, preview].
function capabilityFor(toolName, toolInput) {
  const ti = toolInput || {};

  if (toolName.startsWith('mcp__')) {
    const tool = toolName.split('__').slice(2).join('__').toLowerCase() || toolName.toLowerCase();
    if (/delete|remove|drop|destroy/.test(tool)) return [['mcp:delete', `mcp:${tool}`], toolName];
    if (/send|email|post|publish|transfer|pay/.test(tool)) return [['mcp:send', `mcp:${tool}`], toolName];
    if (/write|create|update|edit|set/.test(tool)) return [['mcp:write', `mcp:${tool}`], toolName];
    return [['mcp:read', `mcp:${tool}`], toolName];
  }

  if (['Edit', 'Write', 'MultiEdit', 'NotebookEdit'].includes(toolName)) {
    const path = String(ti.file_path || '');
    const content = ti.new_string || ti.file_text || ti.content || '';
    return [writeCapsForPath(path), String(content).slice(0, 4000)];
  }

  if (toolName === 'Read') {
    const path = String(ti.file_path || '');
    return [SECRET_PATH.test(path) ? ['secrets:read'] : ['files:read'], path];
  }

  if (toolName === 'WebFetch' || toolName === 'WebSearch') {
    return [['network:http:read'], String(ti.url || ti.query || '')];
  }

  if (toolName === 'Bash') {
    const cmd = String(ti.command || '');
    const caps = shellCapabilities(cmd);
    if (caps && caps.length) return [caps, cmd];
    return [['shell:exec'], cmd]; // parsed clean OR unparseable → opaque exec
  }

  return [[`tool:${toolName.toLowerCase()}`], JSON.stringify(ti).slice(0, 2000)];
}

function decide(toolName, toolInput, profile, mode) {
  const [caps] = capabilityFor(toolName, toolInput);
  return resolve(caps, profile, mode);
}

// Resolve capability candidates against a profile+mode → {decision, reason,
// capability, would}. `would` is the mode-independent verdict, so observe can
// report a would-be block. Shared by the hook and check.
function resolve(caps, profile, mode) {
  const blocked = caps.find((c) => profile.blocks(c));
  const asked = caps.find((c) => profile.asks(c));
  const floor = caps.find((c) => alwaysDenied(c));

  let would;
  let cap;
  if (blocked || floor) { would = 'deny'; cap = floor || blocked; } else if (asked) { would = 'ask'; cap = asked; } else {
    return { decision: 'allow', would: 'allow', capability: caps[0] || '?', reason: '' };
  }

  const pname = profile.name;

  if (mode === 'observe') {
    const verb = would === 'deny' ? 'blocked' : 'held for approval';
    return { decision: 'allow', would, capability: cap, reason: `[observe] would have ${verb} '${cap}' (${pname}).` };
  }

  if (mode === 'ask') {
    if (floor) {
      return { decision: 'deny', would: 'deny', capability: floor, reason: `Certior · ${pname}: '${floor}' is never allowed for an agent.` };
    }
    return { decision: 'ask', would, capability: cap, reason: `Certior · ${pname}: '${cap}' is risky — approve before it runs.` };
  }

  if (would === 'deny') {
    return { decision: 'deny', would: 'deny', capability: cap, reason: `Certior · ${pname}: '${cap}' is outside this boundary.` };
  }
  return { decision: 'ask', would: 'ask', capability: cap, reason: `Certior · ${pname}: '${cap}' is high-stakes — approve before it runs.` };
}

module.exports = { capabilityFor, decide, resolve };

File: certior_guard/engine.py
"""Map a Claude Code tool call to a capability, then decide it against a profile.

The decision logic is deliberately small and dependency-free:

    tool call  ──capability_for──▶  capability strings  ──decide──▶  allow / ask / deny

``capability_for`` translates a PreToolUse envelope (``Bash`` command, ``Edit``
path, ``mcp__…`` call …) into candidate capability strings. ``decide`` matches
those against the chosen :class:`~certior_guard.profiles.Profile` under the
active mode and returns a verdict with a human reason.
"""
from __future__ import annotations

import json
import re
from typing import Any, Dict, List, Tuple

from certior_guard.profiles import Profile, always_denied
from certior_guard.shell_parse import shell_capabilities

# ── Path heuristics ──────────────────────────────────────────────────────────
_SECRET_PATH = re.compile(
    r"(^|/)\.env(\.|$)|(^|/)(secrets?|credentials?)(/|\.|$)|\.pem$|\.key$|"
    r"id_rsa|/\.aws/|/\.ssh/|\.pfx$|\.p12$|\.netrc|service-account\.json|\.npmrc$|\.pypirc$",
    re.IGNORECASE,
)
_PROD_PATH = re.compile(
    r"(^|/)(prod|production|infra|terraform|k8s|kubernetes|helm)(/|$)|"
    r"(^|/)(Dockerfile|docker-compose\.ya?ml)$",
    re.IGNORECASE,
)
_MIGRATION_PATH = re.compile(r"(^|/)(migrations?)(/|$)|/migrate/", re.IGNORECASE)
_AUTH_PATH = re.compile(r"(^|/)(auth|permissions?|security|crypto)(/|$)", re.IGNORECASE)
_BILLING_PATH = re.compile(r"(^|/)(billing|payments?|stripe)(/|$)", re.IGNORECASE)
_CI_PATH = re.compile(r"(^|/)\.github/workflows/|(^|/)\.gitlab-ci\.ya?ml$|(^|/)\.circleci/", re.IGNORECASE)


def _write_caps_for_path(path: str) -> List[str]:
    """Capability candidates for a file write, most-specific first."""
    if _SECRET_PATH.search(path):
        return ["secrets:write", "files:write"]
    if _MIGRATION_PATH.search(path):
        return ["migrations:write", "files:write"]
    if _CI_PATH.search(path):
        return ["ci:write", "files:write"]
    if _AUTH_PATH.search(path):
        return ["auth:write", "files:write"]
    if _BILLING_PATH.search(path):
        return ["billing:write", "files:write"]
    if _PROD_PATH.search(path):
        return ["prod:write", "files:write"]
    return ["files:write"]


def capability_for(tool_name: str, tool_input: Dict[str, Any]) -> Tuple[List[str], str]:
    """Return ``(capability_candidates, preview)`` for a Claude Code tool call.

    Several candidates are returned so a profile glob (``secrets:*``, ``*:write``)
    can match on any of them.
    """
    ti = tool_input or {}

    if tool_name.startswith("mcp__"):
        # mcp__<server>__<tool> — no upstream schema here, so classify by verb.
        tool = tool_name.split("__", 2)[-1].lower()
        if re.search(r"delete|remove|drop|destroy", tool):
            return ["mcp:delete", "mcp:" + tool], tool_name
        if re.search(r"send|email|post|publish|transfer|pay", tool):
            return ["mcp:send", "mcp:" + tool], tool_name
        if re.search(r"write|create|update|edit|set", tool):
            return ["mcp:write", "mcp:" + tool], tool_name
        return ["mcp:read", "mcp:" + tool], tool_name

    if tool_name in ("Edit", "Write", "MultiEdit", "NotebookEdit"):
        path = str(ti.get("file_path", ""))
        content = ti.get("new_string") or ti.get("file_text") or ti.get("content") or ""
        return _write_caps_for_path(path), str(content)[:4000]

    if tool_name == "Read":
        path = str(ti.get("file_path", ""))
        return (["secrets:read"] if _SECRET_PATH.search(path) else ["files:read"]), path

    if tool_name in ("WebFetch", "WebSearch"):
        return ["network:http:read"], str(ti.get("url") or ti.get("query") or "")

    if tool_name == "Bash":
        cmd = str(ti.get("command", ""))
        caps = shell_capabilities(cmd)
        if caps:
            return caps, cmd
        if caps is not None:
            return ["shell:exec"], cmd   # parsed fine, nothing dangerous
        return ["shell:exec"], cmd       # parse failed → still allow bare exec

    return ["tool:" + tool_name.lower()], json.dumps(ti)[:2000]


# ── Decision ─────────────────────────────────────────────────────────────────

def decide(tool_name: str, tool_input: Dict[str, Any], profile: Profile, mode: str) -> Dict[str, str]:
    """Decide a tool call. Returns ``{decision, reason, capability, would}``.

    ``decision`` is what Certior tells Claude Code now (``allow``/``ask``/``deny``).
    ``would`` is what the rules say regardless of mode — so ``observe`` mode can
    report "would have blocked" without actually interrupting.
    """
    caps, _preview = capability_for(tool_name, tool_input)

    blocked = next((c for c in caps if profile.blocks(c)), None)
    asked = next((c for c in caps if profile.asks(c)), None)
    floor = next((c for c in caps if always_denied(c)), None)

    if blocked or floor:
        would, cap = "deny", (floor or blocked)
    elif asked:
        would, cap = "ask", asked
    else:
        return {"decision": "allow", "would": "allow", "capability": caps[0] if caps else "?", "reason": ""}

    pname = profile.name

    if mode == "observe":
        verb = "blocked" if would == "deny" else "held for approval"
        return {"decision": "allow", "would": would, "capability": cap,
                "reason": f"[observe] would have {verb} '{cap}' ({pname})."}

    if mode == "ask":
        # Catastrophic floor stays hard-denied; everything else just asks.
        if floor:
            return {"decision": "deny", "would": "deny", "capability": floor,
                    "reason": f"Certior · {pname}: '{floor}' is never allowed for an agent."}
        return {"decision": "ask", "would": would, "capability": cap,
                "reason": f"Certior · {pname}: '{cap}' is risky — approve before it runs."}

    # enforce
    if would == "deny":
        return {"decision": "deny", "would": "deny", "capability": cap,
                "reason": f"Certior · {pname}: '{cap}' is outside this boundary."}
    return {"decision": "ask", "would": "ask", "capability": cap,
            "reason": f"Certior · {pname}: '{cap}' is high-stakes — approve before it runs."}
